### PR TITLE
Epmrpp 33462 launches comparison chart

### PR DESCRIPTION
**Pull request by PavelZaletski**
Launches comparison chart

### CODE REVIEW

**Comment by ArtsiomYeliseyenka** [comment]
1) Legend clicks does not work
2) Issues with widget frame resizing.

**Comment by PavelZaletski** [comment]
> 1. Legend clicks does not work
> 2. Issues with widget frame resizing.
> 3. Incorrect redirect links(on clicking widget segments. Discuss format with @DiscoElevator )

#1 and #2 have been fixed. Working on #3.

**Comment by PavelZaletski** [comment]
Issue 3 has been fixed

File: app/src/pages/inside/dashboardItemPage/newWidgets/c3chart/c3chart.jsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  componentWillUnmount() {
+    this.destroyChart();
+  }
+
+  destroyChart() {
+    try {
+      this.chart = this.chart.destroy();
+    } catch (err) {
+      throw new Error('Internal C3 error', err);
+    }
```
Comment (ArtsiomYeliseyenka) [inline app/src/pages/inside/dashboardItemPage/newWidgets/c3chart/c3chart.jsx]: Maybe will be better to use something like this.chart && this.chart.destroy()?

File: app/src/pages/inside/dashboardItemPage/newWidgets/chartConfigs/chartConfigs.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    automationBug: '#f7d63e',
+    automation_bug: '#f7d63e',
+    duration: '#507fd5',
+    failed: '#f36c4a',
+    interrupted: '#a94442',
+    invalid: '#ff3222',
+    investigated: '#87b87f',
+    negative: '#e52e2e',
+    noDefect: '#777777',
+    no_defect: '#777777',
```
Comment (ArtsiomYeliseyenka) [inline app/src/pages/inside/dashboardItemPage/newWidgets/chartConfigs/chartConfigs.js]: Do we really need to duplicate colors and store both cases?

Comment (ArtsiomYeliseyenka) [reply app/src/pages/inside/dashboardItemPage/newWidgets/chartConfigs/chartConfigs.js]: And majority of these colors already defined at `common/constants/` folder, it's will be better to use them.

File: app/src/pages/inside/dashboardItemPage/newWidgets/charts/launchesComparisonChart/launchesComparisonChart.jsx [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+  onChartCreated = (chart, element) => {
+    this.chart = chart;
+    this.node = element;
+
+    if (!this.props.widget.content.result || this.props.isPreview) {
+      return;
+    }
+
+    const legendClassName = cx('legend');
```
Comment (ArtsiomYeliseyenka) [inline app/src/pages/inside/dashboardItemPage/newWidgets/charts/launchesComparisonChart/launchesComparisonChart.jsx]: Create `setupLegend` method, witch will contain legend-related stuff.

Comment (ArtsiomYeliseyenka) [reply app/src/pages/inside/dashboardItemPage/newWidgets/charts/launchesComparisonChart/launchesComparisonChart.jsx]: Moreover, legend should be common reusable, configurable component which can be used in others widgets. Please think about moving it outside.

File: app/src/pages/inside/dashboardItemPage/newWidgets/charts/launchesComparisonChart/launchesComparisonChart.scss [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+  &:before {
+    position: absolute;
+    display: none;
+    left: 2px;
+    top: 2px;
+    width: 8px;
+    height: 8px;
+    border-radius: 100%;
+    background-color: #fff;
```
Comment (ArtsiomYeliseyenka) [inline app/src/pages/inside/dashboardItemPage/newWidgets/charts/launchesComparisonChart/launchesComparisonChart.scss]: It's not allowed to define colors by hex codes. Please use existing color variables.

Comment (ArtsiomYeliseyenka) [reply app/src/pages/inside/dashboardItemPage/newWidgets/charts/launchesComparisonChart/launchesComparisonChart.scss]: Check for other cases.

File: app/src/pages/inside/dashboardItemPage/newWidgets/charts/launchesComparisonChart/launchesComparisonChart.scss [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+.legend-item {
+  cursor: pointer;
+  margin-right: 15px;
+}
+
+:global {
+  .c3-grid .c3-ygrid {
+    stroke-dasharray: none;
+    stroke: #ccc;
```
Comment (ArtsiomYeliseyenka) [inline app/src/pages/inside/dashboardItemPage/newWidgets/charts/launchesComparisonChart/launchesComparisonChart.scss]: fix

File: app/src/pages/inside/dashboardItemPage/newWidgets/tooltip/tooltip.jsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        <div className={cx('launch-name')}>{`${launchName} #${launchNumber}`}</div>
+        <div className={cx('launch-start-time')}>{this.dateFormat(startTime)}</div>
+        <div className={cx('item-wrapper')}>
+          <span className={cx('color-mark')} style={{ backgroundColor: color }} />
+          <span className={cx('item-name')}>{`${itemName}:`}</span>
+          <span className={cx('item-cases')}>
+            <span>{`${itemCases}%`}</span>
+          </span>
+        </div>
+      </div>
```
Comment (ArtsiomYeliseyenka) [inline app/src/pages/inside/dashboardItemPage/newWidgets/tooltip/tooltip.jsx]: Tooltip content can be different for different widget types. It's good solution to make a common tooltip component for all widgets. But it's required to make ability to provide different content to that.

File: app/src/pages/inside/dashboardItemPage/newWidgets/tooltip/tooltip.scss [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+.launch-start-time {
+  margin-bottom: 4px;
+}
+
+.item-name {
+  margin-right: 4px;
+}
+
+.item-cases {
+  font-weight: bold;
```
Comment (ArtsiomYeliseyenka) [inline app/src/pages/inside/dashboardItemPage/newWidgets/tooltip/tooltip.scss]: Forbidden css property. Use corresponding font variable

File: app/src/pages/inside/dashboardItemPage/newWidgets/charts/launchesComparisonChart/launchesComparisonChart.scss (hunk 4 above)
Comment (PavelZaletski) [reply app/src/pages/inside/dashboardItemPage/newWidgets/charts/launchesComparisonChart/launchesComparisonChart.scss]: Fixed

File: app/src/pages/inside/dashboardItemPage/newWidgets/charts/launchesComparisonChart/launchesComparisonChart.scss (hunk 5 above)
Comment (PavelZaletski) [reply app/src/pages/inside/dashboardItemPage/newWidgets/charts/launchesComparisonChart/launchesComparisonChart.scss]: Fixed

File: app/src/pages/inside/dashboardItemPage/newWidgets/tooltip/tooltip.scss (hunk 7 above)
Comment (PavelZaletski) [reply app/src/pages/inside/dashboardItemPage/newWidgets/tooltip/tooltip.scss]: Fixed

File: app/src/pages/inside/dashboardItemPage/newWidgets/c3chart/c3chart.jsx (hunk 1 above)
Comment (PavelZaletski) [reply app/src/pages/inside/dashboardItemPage/newWidgets/c3chart/c3chart.jsx]: Done

File: app/src/pages/inside/dashboardItemPage/newWidgets/chartConfigs/chartConfigs.js (hunk 2 above)
Comment (PavelZaletski) [reply app/src/pages/inside/dashboardItemPage/newWidgets/chartConfigs/chartConfigs.js]: Fixed. Moved colors to common file

File: app/src/pages/inside/dashboardItemPage/charts/common/c3chart/c3chart.jsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    onChartCreated: () => {},
+    className: '',
+    style: {},
+  };
+
+  componentDidMount() {
+    this.updateChart(this.props.config);
+  }
+
+  componentWillReceiveProps(newProps) {
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/charts/common/c3chart/c3chart.jsx]: `componentDidUpdate`

https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html

File: app/src/pages/inside/dashboardItemPage/charts/common/c3chart/c3chart.jsx [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  render() {
+    const className = this.props.className ? ` ${this.props.className}` : '';
+    const style = this.props.style ? this.props.style : {};
+    return (
+      <React.Fragment>
+        <div
+          ref={(node) => {
+            this.node = node;
+          }}
+          className={className}
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/charts/common/c3chart/c3chart.jsx]:
```
className={this.props.className}
style={this.props.style}
```

File: app/localization/translated/be.json [hunk 10]
```diff
@@ -256,6 +256,14 @@
   "Filter.edit": "Рэдагаваць фільтр",
   "Filter.name": "Назва",
   "Filter.namePlaceholder": "Увядзіце назву фільтра",
+  "filterNameById.statistics$defects$automation_bug": "Хіба аўтаматызацыі",
```
Comment (DiscoElevator) [inline app/localization/translated/be.json]: FilterNameById

File: app/src/pages/inside/dashboardItemPage/charts/common/tooltip/renderToHtml.js [hunk 11]
```diff
@@ -0,0 +1,7 @@
+import { render } from 'react-dom';
+
+export function renderToHtml(component) {
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/charts/common/tooltip/renderToHtml.js]: You may use this function:
```
import ReactDOMServer from 'react-dom/server';

ReactDOMServer.renderToStaticMarkup(//...)
```

File: app/src/pages/inside/dashboardItemPage/charts/common/tooltip/tooltipContent/tooltipContent.jsx [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+  static defaultProps = {
+    launchName: '',
+    launchNumber: '',
+    startTime: '',
+    itemCases: 0,
+    itemName: '',
+  };
+
+  dateFormat = (val) => {
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/charts/common/tooltip/tooltipContent/tooltipContent.jsx]: Use `dateFormat` from `common/utils/timeDateUtils.js`

File: app/src/pages/inside/dashboardItemPage/charts/launchesComparisonChart/launchesComparisonChart.jsx [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+          itemCases={d[0].value}
+          color={color(id)}
+          itemName={this.props.intl.formatMessage(messages[id.split('$total')[0]])}
+        />
+      </TooltipWrapper>,
+    );
+  };
+
+  render() {
+    this.getData();
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/charts/launchesComparisonChart/launchesComparisonChart.jsx]: Why it's in the render function?

File: app/src/pages/inside/dashboardItemPage/charts/launchesComparisonChart/launchesComparisonChart.jsx [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+        height: newHeight,
+      });
+      this.height = newHeight;
+    } else if (this.width !== newWidth) {
+      this.chart.flush();
+      this.width = newWidth;
+    }
+  };
+
+  contents = (d, defaultTitleFormat, defaultValueFormat, color) => {
```
Comment (DiscoElevator) [inline app/src/pages/inside/dashboardItemPage/charts/launchesComparisonChart/launchesComparisonChart.jsx]: renderContents

File: app/package.json [hunk 15]
```diff
@@ -45,6 +48,7 @@
     "react-datepicker": "^1.4.1",
     "react-dom": "^16.3.2",
     "react-dropzone": "^4.2.9",
+    "react-event-observer": "^0.5.11",
```
Comment (DiscoElevator) [inline app/package.json]: What is it for?

File: app/src/pages/inside/dashboardItemPage/charts/common/c3chart/c3chart.jsx (hunk 8 above)
Comment (PavelZaletski) [reply app/src/pages/inside/dashboardItemPage/charts/common/c3chart/c3chart.jsx]: Fixed

File: app/src/pages/inside/dashboardItemPage/charts/common/c3chart/c3chart.jsx (hunk 9 above)
Comment (PavelZaletski) [reply app/src/pages/inside/dashboardItemPage/charts/common/c3chart/c3chart.jsx]: Fixed

File: app/src/pages/inside/dashboardItemPage/charts/launchesComparisonChart/launchesComparisonChart.jsx (hunk 14 above)
Comment (PavelZaletski) [reply app/src/pages/inside/dashboardItemPage/charts/launchesComparisonChart/launchesComparisonChart.jsx]: Renamed

**Comment by PavelZaletski** [comment]
closed in favor of #981

**Pull request closed without merging**